File: golang/1481.least-number-of-unique-integers-after-k-removals.go
func findLeastNumOfUniqueInts(arr []int, k int) int {
    freq := make(map[int]int)
    for _, num := range arr { freq[num]++ }

    res := len(freq)
    n := len(arr)
    quef := make([]int, n + 1)

    for _, f := range freq {
        quef[f]++
    }

    for f := 1; f <= n; f++ {
        if k >= quef[f] * f {
            k -= quef[f] * f
            res -= quef[f]
        } else {
            res -= k / f
            break
        }
    }

    return res
}

File: cpp/1481.least-number-of-unique-integers-after-k-removals.cpp
class Solution {
public:
    int findLeastNumOfUniqueInts(vector<int>& arr, int k) {
        map<int, int> ch;

        for (auto num: arr) {
            ch[num]++;
        }

        priority_queue<int, vector<int>, std::greater<int>> hq;
        
        for (auto [_, v]: ch) {
            hq.push(v);
        }

        int res = hq.size();

        while (k > 0 && !hq.empty() && hq.top() <= k) {
            k -= hq.top(); hq.pop();
            res--;
        }

        return res;
    }
};